### PR TITLE
Insights 12

### PR DESCRIPTION
**Pull request by robert-close**
Adding method to allow users to send insight information from a python algorithm to the queue reader. I am not very experienced with Python, so please point out any areas for improvement or correction. This is a pretty simple flow and I could have just called the request.post from the client, but I feel it is cleaner to have the Insights class segregated.

### CODE REVIEW

**Comment by robert-close** [comment]
These test failures are not clear to me. Is it a 2.7 problem?

**Comment by robert-close** [comment]
Should I create a Insight object to enforce the insightKey and InsightValue field names?

File: Algorithmia/insights.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import requests
+import json
+
+
+class Insights:
+    # Example of correct insights:
+    # {
```
Comment (dinoboy197) [inline Algorithmia/insights.py:7]: Would it be possible to use a more streamlined key / value pair structure? For example, what we've written up in [the documentation proposal for Python](https://insights1.enthalpy.click/developers/clients/python#publishing-algorithmia-insights) so far? (Sorry I didn't get this to you last week):
```
client.report_insights({"cats_in_image": 4, "dogs_in_image": 7})
```

Comment (robert-close) [reply Algorithmia/insights.py:7]: Got it, I had not seen that document before. I changed the method signatures as well

File: Algorithmia/insights.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    #   {"aKey":"aValue"},
+    #   {"aKey2":"aValue2"}
+    # }
+    def __init__(self, insights):
+        # TODO should we get the URL from a config?
```
Comment (kennydaniel) [inline Algorithmia/insights.py:12]: I would like to see this use an environment variable with a fallback to localhost. First of all, when doing local dev, 9000 will often already be taken by play server (but this is expecting to hit a sidecar right?). Also, it's difficult to get users to update python client on a schedule, so having more flexibility from the start is valuable.

File: Test/algo_test.py [hunk 3]
```diff
@@ -11,10 +11,11 @@ class AlgoTest(unittest.TestCase):
     def setUp(self):
         self.client = Algorithmia.client()
 
-    def test_call_binary(self):
-        result = self.client.algo('util/Echo').pipe(bytearray('foo','utf-8'))
-        self.assertEquals('binary', result.metadata.content_type)
-        self.assertEquals(bytearray('foo','utf-8'), result.result)
+# todo fix this
```
Comment (kennydaniel) [inline Test/algo_test.py:14]: This seems unrelated to the insights code, why did the test need to be commented out?

Comment (robert-close) [reply Test/algo_test.py:14]: This test is failing in the pipeline, I am not sure what is causing it.

File: Algorithmia/insights.py (hunk 2 above)
Comment (robert-close) [reply Algorithmia/insights.py]: I created this Jira to set the env variable. https://algorithmia.atlassian.net/browse/INSIGHTS-36. This code has been updated to default to the localhost:9000 if it is not set.

**Comment by robert-close** [comment]
The build is failing for an unrelated test. Any help would be appreciated.

**Comment by dinoboy197** [comment]
@kennydaniel [the failure in the tests](https://github.com/algorithmiaio/algorithmia-python/runs/1098200923) seems to be:
```
__________________________ AlgoTest.test_call_binary ___________________________

self = <Test.algo_test.AlgoTest testMethod=test_call_binary>

    def test_call_binary(self):
        result = self.client.algo('util/Echo').pipe(bytearray('foo','utf-8'))
        self.assertEquals('binary', result.metadata.content_type)
>       self.assertEquals(bytearray('foo','utf-8'), result.result)
E       AssertionError: bytearray(b'foo') != 'Zm9v'
```

Given that you and Ben worked on this recently, would you have any idea what this might be caused by?

File: setup.py [hunk 4]
```diff
@@ -1,10 +1,11 @@
+import distutils
 import os
 
 from setuptools import setup
 
 setup(
     name='algorithmia',
-    version='1.4.0',
+    version='1.4.4',
```
Comment (dinoboy197) [inline setup.py:8]: Since we're adding new functionality, it seems like a minor version bump (`1.4.0` -> `1.5.0`) would be more appropriate given semver - thoughts @ZachEddy ?

File: setup.py [hunk 5]
```diff
@@ -1,10 +1,11 @@
+import distutils
```
Comment (dinoboy197) [inline setup.py:1]: Is `distutils` a new library that is needed for setup but somehow wasn't before?

**Review by dinoboy197 — COMMENTED** [review]
I'm curious about that failing test; I can't see how it's related to your code, but I'm hoping Kenny might have a thought on that?

File: setup.py (hunk 4 above)
Comment (ZachEddy) [reply setup.py:8]: Yep, I agree. Bumping the minor version makes sense here.

Sidenote: weirdly, this is the only Slack notification that GitHub sent me for this pull request, so I've been missing tons of activity 😕

File: setup.py (hunk 5 above)
Comment (robert-close) [reply setup.py]: Sorry, that was a left over from an unneeded change I tried.

File: setup.py (hunk 4 above)
Comment (robert-close) [reply setup.py]: I wanted to wait to do that bump until we are fully tested as we can only use 1.5.0 once in Pypi, but I will add this to the commit here.

File: Algorithmia/insights.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+class Insights:
+    # Example of correct insights:
+    # {"aKey":"aValue","aKey2":"aValue2"}
+    def __init__(self, insights):
+        headers = {}
+        headers['Content-Type'] = 'application/json'
+        AQR_URL = os.getenv('ALGORITHM_QUEUE_READER_URL') or "http://localhost:9000"
+
+        requests.post(AQR_URL+"/insights", data=json.dumps(insights).encode('utf-8'), headers=headers)
```
Comment (ZachEddy) [inline Algorithmia/insights.py:13]: Something that came out of the design review: the endpoint in AQR should be prefixed with `/v1/` so we can safely make breaking API changes without breaking the `v1.X.X` client code.

I still need to modify AQR and add the prefix but that is a small change, so let's go ahead and make the change in this PR.

```
/insights --> /v1/insights
```

File: Algorithmia/client.py [hunk 7]
```diff
@@ -40,6 +41,11 @@ def dir(self, dataUrl):
         if dataUrl.startswith('file://'): return LocalDataDirectory(self, dataUrl)
         else: return DataDirectory(self, dataUrl)
 
+    # Used to send insight data to Algorithm Queue Reader in cluster
+    def report_insights(self, insights):
+        return Insights(insights)
```
Comment (kennydaniel) [inline Algorithmia/client.py:46]: Two comments:
1) why are we returning theInsights object? Seems like void return type makes more sense here.
2) I think this should be wrapped in a try/catch so make local running easier. Right now I think it will throw exception if running locally with insights.

**Comment by dinoboy197** [comment]
Will rebuild this PR once [AQR with binary payload fix](https://algorithmia.atlassian.net/browse/PSE-326) is deployed to production public marketplace.

**Comment by dinoboy197** [comment]
@algorithmiaio/developer-technologies-contractors ping for a review please!

**Review by John-Bragg — APPROVED** [review]

**Comment by dinoboy197** [comment]
Thanks for your approval @John-Bragg ! You can just leave the PRs open once you approve them because we use Github to merge the changes.

**Comment by dinoboy197** [comment]
@robert-close ready for you to merge!
